paper.tex

```diff
@@ -15,6 +15,77 @@ \section{Architecture}
 %% I volunteer to write this section. --Aaron
 %%
 %% Representing symbolic expressions using Python objects
+
+Every symbolic expression in SymPy is an instance of a Python class.
+Expressions are represented by expression trees. The operators is represented
+by the type of an expression and the child nodes are stored in the
+\texttt{args} attribute. A leaf node in the expression tree has an empty
+\texttt{args}. For example, take the expression $xy + 2$.
+
+\begin{verbatim}
+>>> from sympy import *
+>>> x, y = symbols('x y')
+>>> expr = x*y + 2
+\end{verbatim}
+
+The expression \texttt{expr} is an addition, so it is of type \texttt{Add}. The child
+nodes of \texttt{expr} are \texttt{x*y} and \texttt{2}.
+
+\begin{verbatim}
+>>> type(expr)
+<class 'sympy.core.add.Add'>
+>>> expr.args
+(2, x*y)
+\end{verbatim}
+
+We can dig further into the expression tree to see the full expression. For
+example, the first child node, given by \texttt{expr.args[0]} is
+\texttt{2}. Its class is \texttt{Integer}, and it has empty \texttt{args},
+indicating that it is a leaf node
+
+\begin{verbatim}
+>>> expr.args[0]
+2
+>>> type(expr.args[0])
+<class 'sympy.core.numbers.Integer'>
+>>> expr.args[0].args
+()
+\end{verbatim}
+
+Every SymPy expression satisfies a key invariant, namely,
+\texttt{expr.func(*expr.args) == expr}. This means that expressions are
+rebuildable from their \texttt{args}\footnote{\texttt{expr.func} is used
+  instead of \texttt{type(expr)} to allow the function of an expression to be
+  distinct from its actual Python class. In most cases the two are the same.}.
+Here, we note that in SymPy, the \texttt{==} operator represents exact
+structural equality, not mathematical equality. This allows us to test if any
+two expressions are equal to one another as expression trees.
+
+Python allows classes to overload operators. The Python interpreter translates
+above \texttt{x*y + 2} to, roughly,
+\texttt{(x.\_\_mul\_\_(y)).\_\_add\_\_(2)}. \texttt{x} and \texttt{y}, returned from
+the \texttt{symbols} function, are \texttt{Symbol} instances. The \texttt{2}
+in the expression is processed by Python as a literal, and is stored as
```

Comment:
The point I want to get across here is that even though the `2` is interpreted by Python as an `int`, it gets converted to Integer when it gets passed in to the arguments of `__add__`.